src/vendor/azure-uamqp-c/src/session.c

```diff
@@ -862,15 +862,30 @@ int session_begin(SESSION_HANDLE session)
         {
             if (!session_instance->is_underlying_connection_open)
             {
-                if (connection_open(session_instance->connection) != 0)
+                CONNECTION_OPEN_RESULT connection_open_result;
+                connection_open_result = connection_open(session_instance->connection);
+                if (connection_open_result == CONNECTION_OPEN_OK)
                 {
-                    session_instance->is_underlying_connection_open = UNDERLYING_CONNECTION_NOT_OPEN;
-                    result = __FAILURE__;
+                    session_instance->is_underlying_connection_open = UNDERLYING_CONNECTION_OPEN;
+                    result = 0;
                 }
-                else
+                else if(connection_open_result == CONNECTION_OPEN_ALREADY_OPEN)
                 {
                     session_instance->is_underlying_connection_open = UNDERLYING_CONNECTION_OPEN;
-                    result = 0;
+                    if (send_begin(session) != 0)
```

Comment:
Why do we only call `send_begin` if `connection_open_result == CONNECTION_OPEN_ALREADY_OPEN` and not if `connection_open_result == CONNECTION_OPEN_OK`?

Reply:
The `send_begin` function can only be called when the connection has been already opened, that is the connection state being `CONNECTION_STATE_OPENED`.

If it's a new session on new connection, the `send_begin` will be called in the [`on_connection_state_changed`](https://github.com/Azure/azure-uamqp-python/blob/0a631b33f46ccbea25ffa15d811e5eb3a4a7edc4/src/vendor/azure-uamqp-c/src/session.c#L363).

However, if it's an existing opened connection, there is no chance of running into the `on_connection_state_changed` and newew session couldn't be begun - that's the reason we previously don't support multiple sessions on one connection and why we only need the `session_begin` for newer sessions in `CONNECTION_OPEN_ALREADY_OPEN` case.

But overall the `connection_open_result` is just an indicator for the creation result of the connection object. It still can't indicate the connection_state, which means the `send_begin` could fail if a connection has been created but too hurry to finish the processing of opening(shaking hands with the server side).

There is certain situation that multiple sessions are started on one single connection simultaneously, the relatively-latter session would check and find connection already created while the underlying connection has not finished shaking hands. So this would lead to session begin failure.

To handle the situation I described above, I think it's the application-level responsibility to ensure the connection is opened before we want to multiplexing it. So I put extra logic in the context manager and `redirect` of connection to get it opened first.

Well, this is the whole story of enabling multiple sessions. Hope it won't confuse you a lot.

Let me know if you have more questions.